extensions/src/main/java/org/pentaho/platform/spring/ApplicationEventDispatcher.java

```diff
@@ -0,0 +1,140 @@
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+
+public class ApplicationEventDispatcher extends AbstractApplicationEventMulticaster {
+  public final static String ASYNC_ANNOTATION_QUALIFIER = "dispatcher";
+
+  private ApplicationEventMulticaster asyncMulticaster;
+  private ApplicationEventMulticaster defaultMulticaster;
+
+  private BeanFactory beanFactory;
+
+  @Override
+  public void addApplicationListener( final ApplicationListener listener ) {
+    if ( isAsync( listener ) ) {
+      asyncMulticaster.addApplicationListener( listener );
+    } else {
+      defaultMulticaster.addApplicationListener( listener );
+    }
+  }
+
+  @Override
+  public void removeApplicationListener( final ApplicationListener listener ) {
+    asyncMulticaster.removeApplicationListener( listener );
+    defaultMulticaster.removeApplicationListener( listener );
+  }
+
+  @Override
+  public void removeApplicationListenerBean( final String listenerBeanName ) {
+    asyncMulticaster.removeApplicationListenerBean( listenerBeanName );
+    defaultMulticaster.removeApplicationListenerBean( listenerBeanName );
+  }
+
+  @Override
+  public void removeAllListeners() {
+    defaultMulticaster.removeAllListeners();
+    asyncMulticaster.removeAllListeners();
+  }
+
+  public void multicastEvent( final ApplicationEvent event ) {
+    defaultMulticaster.multicastEvent( event );
+    asyncMulticaster.multicastEvent( event );
+  }
+
+  public void multicastEvent(final ApplicationEvent event, final ResolvableType type ) {
+    defaultMulticaster.multicastEvent( event, type );
+    asyncMulticaster.multicastEvent( event, type );
+  }
+
+  public void setAsyncMulticaster( final ApplicationEventMulticaster asyncMulticaster ) {
+    this.asyncMulticaster = asyncMulticaster;
+  }
+
+  public void setDefaultMulticaster( final ApplicationEventMulticaster defaultMulticaster ) {
+    this.defaultMulticaster = defaultMulticaster;
+  }
+
+  @Override
+  public void setBeanFactory( final BeanFactory beanFactory ) {
+    super.setBeanFactory( beanFactory );
+
+    if ( defaultMulticaster instanceof AbstractApplicationEventMulticaster ) {
+      ( (AbstractApplicationEventMulticaster) defaultMulticaster ).setBeanFactory( beanFactory );
+    }
+
+    if ( asyncMulticaster instanceof AbstractApplicationEventMulticaster ) {
+      ( (AbstractApplicationEventMulticaster) asyncMulticaster ).setBeanFactory( beanFactory );
+    }
+
+    this.beanFactory = beanFactory;
+  }
+
+  @Override
+  public void addApplicationListenerBean( final String listenerBeanName ) {
+    final ApplicationListener listener = beanFactory.getBean( listenerBeanName, ApplicationListener.class );
+    if ( isAsync( listener ) ) {
+      asyncMulticaster.addApplicationListenerBean( listenerBeanName );
+    } else {
+      defaultMulticaster.addApplicationListenerBean( listenerBeanName );
+    }
+  }
+
+  private boolean isAsync( final ApplicationListener listener ) {
```

Comment:
Well yes.
 
However the whole point was that  spring did not process the Async annotation properly (it simply did not work) - it used a multicaster class that could do either/or but not both type of dispatching. To achieve the ability to use both sync and async event dispatching this was one of the solutions suggested online.
(ApplicationEventMulticaster can be configured with either - sync/async task executor)